content/posts/2020/02/thai-in-2562/index.eng.md

```diff
@@ -0,0 +1,93 @@
+
+Even though the method may be non-deterministic, the API designed to follow the locale of the execution environment is probably intended because such formatting operations are usually used for user interface, and coding can naturally look appropriate to the cultural community without much concern about internationalization. But the reason why we used this method was not for user interface, but for the cryptographic hash input that had to be deterministic— this turned out to be a mistake.
+
+Now that we know the cause, we have solved the urgent problem by [finding a method that has the `CultureInfo` or `IFormatProvider` parameters omitted, and patching it to explicitly designate `CultureInfo.InvariantCulture`, just like the method `DateTimeOffset.ToString`()][libplanet#734].
+
+CI has also been reinforced with unit testing in Arabic, French, Hebrew locale and so on. Since there are a lot of countries in Europe that use comma (`,`) instead of period (`.`) in decimal places, and countries in the Middle East that writes from right to left, we deliberately chose language regions that were somewhat unfamiliar to us.
+
+In addition, because similar mistakes can happen in the future, we have [introduced static analysis][libplanet#737] that finds codes whose behavior depends on the locale of the execution environment.
+
+[1]: https://github.com/planetarium/libplanet/blob/82aaba0c37591ebf51207038e8c5c122272ce98b/Libplanet/Blocks/Block.cs#L488
+[2]: https://docs.microsoft.com/en-us/dotnet/api/system.datetimeoffset.tostring?view=netstandard-2.0#System_DateTimeOffset_ToString
+[Gregorian calendar]: https://en.wikipedia.org/wiki/Gregorian_calendar
+[Buddhist calendar]: https://en.wikipedia.org/wiki/Buddhist_calendar
+[Thai solar calendar]: https://en.wikipedia.org/wiki/Thai_solar_calendar
+[IFormatProvider]: https://docs.microsoft.com/en-us/dotnet/api/system.iformatprovider?view=netstandard-2.0
+[CultureInfo]: https://docs.microsoft.com/en-us/dotnet/api/system.globalization.cultureinfo?view=netstandard-2.0
+[CultureInfo.CurrentCulture]: https://docs.microsoft.com/en-us/dotnet/api/system.globalization.cultureinfo.currentculture?view=netstandard-2.0
+[CultureInfo.InvariantCulture]: https://docs.microsoft.com/en-us/dotnet/api/system.globalization.cultureinfo.invariantculture?view=netstandard-2.0
+[libplanet#734]: https://github.com/planetarium/libplanet/pull/734
+[libplanet#737]: https://github.com/planetarium/libplanet/pull/737
+
+
+## Closing
+
+As mentioned earlier, using APIs with non-determinant behaviors such as formatting in functions that calculate cryptographic hash is not a good decision in the long run. Typically, because strings are heavily formatted, it's safe to avoid them from data level perspective.
```

Comment:
```suggestion
As mentioned earlier, using APIs with non-determinant behaviors such as formatting in functions that calculate the cryptographic hash is not a good decision in the long run. Typically, because strings are heavily formatted, it's safe to avoid them from data level perspective.
```